### PR TITLE
[Form] Deprecated setDefaultOptions() in favor of configureOptions()

### PR DESCRIPTION
**Pull request by peterrehm**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? |  |
| Fixed tickets | #12782 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4786 |

This tries to provide a compatible API with the depreciation of the OptionResolverInterface. I would like to have this in 2.6.2 but I think that might not be possible? To me I think we should always provide an API where you do not need to use deprecated classes.

Also can you think of any way to trigger errors on the use of the deprecated setDefaultOptions() method? Since it is usually overwritten without calling the parent class this might be tricky. Maybe only in the resolver if we can check if actual options has been resolved in a call to setDefaultOptions.

### CODE REVIEW

**Comment by peterrehm** [comment]
Any opinion on this PR?

File: src/Symfony/Component/Form/AbstractType.php [hunk 1]
```diff
@@ -47,6 +48,15 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
     }
 
     /**
+     * Configure the options for this type.
+     *
+     * @param OptionsResolver $resolver The resolver for the options.
+     */
+    public function configureOptions(OptionsResolver $resolver)
```
Comment (stof) [inline src/Symfony/Component/Form/AbstractType.php]: ``setDefaultOptions`` must call ``configureOptions`` to be BC

File: src/Symfony/Component/Form/AbstractTypeExtension.php [hunk 2]
```diff
@@ -45,4 +46,13 @@ public function finishView(FormView $view, FormInterface $form, array $options)
     public function setDefaultOptions(OptionsResolverInterface $resolver)
     {
     }
+
+    /**
+     * Configure the options for this type.
+     *
+     * @param OptionsResolver $resolver The resolver for the options.
+     */
+    public function configureOptions(OptionsResolver $resolver)
```
Comment (stof) [inline src/Symfony/Component/Form/AbstractTypeExtension.php]: same here

File: src/Symfony/Component/Form/FormTypeExtensionInterface.php [hunk 3]
```diff
@@ -63,6 +63,9 @@ public function finishView(FormView $view, FormInterface $form, array $options);
      * Overrides the default options from the extended type.
      *
      * @param OptionsResolverInterface $resolver The resolver for the options.
+     *
+     * @deprecated Deprecated since Symfony 2.7, to be removed in Symfony 3.0.
+     *             Use {@link FormTypeExtenstionInterface::configureOptions()} instead.
```
Comment (stof) [inline src/Symfony/Component/Form/FormTypeExtensionInterface.php]: the method is not in the interface yet

File: src/Symfony/Component/Form/FormTypeInterface.php [hunk 4]
```diff
@@ -72,6 +72,9 @@ public function finishView(FormView $view, FormInterface $form, array $options);
      * Sets the default options for this type.
      *
      * @param OptionsResolverInterface $resolver The resolver for the options.
+     *
+     * @deprecated Deprecated since Symfony 2.7, to be removed in Symfony 3.0.
+     *             Use {@link FormTypeInterface::configureOptions()} instead.
```
Comment (stof) [inline src/Symfony/Component/Form/FormTypeInterface.php]: the method is not in the interface yet

File: src/Symfony/Component/Form/ResolvedFormType.php [hunk 5]
```diff
@@ -205,8 +205,16 @@ public function getOptionsResolver()
 
             $this->innerType->setDefaultOptions($this->optionsResolver);
 
+            if (method_exists($this->innerType, 'configureOptions')) {
```
Comment (stof) [inline src/Symfony/Component/Form/ResolvedFormType.php]: this ugly check should be remove. Instead, the default implementation of setDefaultOptions in the AbstractType should call the new method

**Comment by stof** [comment]
@fabpot I think this is a valuable change:

- ``setDefaultOptions`` is a weird name given that the method is not only about setting the default options, but about configuring the whole OptionsResolver (the name was probably get close of the Symfony 2.0 ``getDefaultOptions`` even though the responsibility was larger)
- using a different method names allows us to build a forward-compatibility layer for the interface deprecation.

**Comment by fabpot** [comment]
Indeed, this looks like a very good change.

@peterrehm Besides the comments from @stof, you must add the deprecation notices where applicable before we can merge this PR. There are some other things to take care of like moving tests for the old version into "legacy" tests, but I will let @nicolas-grekas comment on this.

File: src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php [hunk 6]
```diff
@@ -103,6 +103,53 @@ public function testGetOptionsResolver()
         $this->assertEquals($resolvedOptions, $resolver->resolve($givenOptions));
     }
 
+    public function testGetOptionsResolverWithConfigureOptions()
+    {
+        if (version_compare(\PHPUnit_Runner_Version::id(), '3.7', '<')) {
+            $this->markTestSkipped('This test requires PHPUnit 3.7.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php]: We removed all those skips recently, please remove this one also.

**Comment by nicolas-grekas** [comment]
As @fabpot said, please trigger deprecations notices when the deprecated methods are used (see #13060).
Please also update any usage of the deprecated methods everywhere in the code base,
please verify also that composer.json files state correct minimum requirements (if the new methods are used cross-components)
last but not least, if tests remain that exercise the deprecated methods, please isolate them in test methods or classes with `testLegacy` (for methods) of `Legacy*Test` (for classes) so that we can easily drop them in 3.0.
I didn't look at the details of your PR, so maybe not all of this is applicable here, but this is the general process.

**Comment by stof** [comment]
to be clear, the case needing to throw a notice is when a type is overwriting setDefaultOptions instead of overwriting configureOptions. The code using the types should still call the ``setDefaultOptions`` method because we cannot change the interface until 3.0 (meaning we don't have a BC layer here, but a forward compatibility layer)

File: src/Symfony/Component/Form/ResolvedFormType.php (hunk 5 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/ResolvedFormType.php]: But this is on purpose to avoid a potential bc break. If someone is not using the abstract type and rather uses a custom implementation of the `FormTypeInterface` we might break applications. Therefore I decided to check if configureOptions() exists and if so calling configureOptions() additionally. I could not come up with a better way. Therefore I did not call configureOptions within the `AbstractType::setDefaultOptions()`.

File: src/Symfony/Component/Form/AbstractType.php (hunk 1 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/AbstractType.php]: It is BC compatible hence the checks for configureOptions in `ResolvedFormType`. See also my comments there.

File: src/Symfony/Component/Form/FormTypeExtensionInterface.php (hunk 3 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/FormTypeExtensionInterface.php]: Yes, I was unsure how to document that. I have rewritten that.

File: src/Symfony/Component/Form/ResolvedFormType.php (hunk 5 above)
Comment (stof) [inline src/Symfony/Component/Form/ResolvedFormType.php]: if someone does not use the AbstractType, they are required to implement setDefaultOptions because of the interface.

IMO, we should stick to the interface for the external usage of the FormTypeInterface, and build the forward-compatibility layer only in AbstractType, by making ``setDefaultOptions`` call ``configureOptions``. This will not break applications

If someone uses the FormTypeInterface without AbstractType, they will have a hard BC break for 3.0 when we will replace the method in the interface, but I haven't seen any code recreating the implementation from scratch rather than extending the base class with empty implementations.

**Comment by peterrehm** [comment]
I have just added a way to check if the method has been overwritten in the `ResolvedFormType`.
That is the only way I can think of, however I do not know about the performance implications of this solution. What do you think about that?

File: src/Symfony/Component/Form/ResolvedFormType.php [hunk 7]
```diff
@@ -205,8 +205,30 @@ public function getOptionsResolver()
 
             $this->innerType->setDefaultOptions($this->optionsResolver);
 
+            $reflector = new \ReflectionMethod($this->innerType, 'setDefaultOptions');
+            $isOverwritten = ($reflector->getDeclaringClass()->getName() !== 'AbstractType');
```
Comment (stof) [inline src/Symfony/Component/Form/ResolvedFormType.php]: This will never work, because the class name is not AbstractType. So it will always be reported as overwritten

File: src/Symfony/Component/Form/ResolvedFormType.php [hunk 8]
```diff
@@ -205,8 +205,30 @@ public function getOptionsResolver()
...
+
+            if (method_exists($this->innerType, 'configureOptions')) {
+                $this->innerType->configureOptions($this->optionsResolver);
+            }
+
             foreach ($this->typeExtensions as $extension) {
                 $extension->setDefaultOptions($this->optionsResolver);
+
+                $reflector = new \ReflectionMethod($extension, 'setDefaultOptions');
+                $isOverwritten = ($reflector->getDeclaringClass()->getName() !== 'AbstractTypeExtension');
```
Comment (stof) [inline src/Symfony/Component/Form/ResolvedFormType.php]: same issue here

**Comment by peterrehm** [comment]
Most likely due to my local squash the comment by @stof got lost:

> if someone does not use the AbstractType, they are required to implement setDefaultOptions because of the interface.

> IMO, we should stick to the interface for the external usage of the FormTypeInterface, and build the forward-compatibility layer only in AbstractType, by making setDefaultOptions call configureOptions. This will not break applications

> If someone uses the FormTypeInterface without AbstractType, they will have a hard BC break for 3.0 when we will replace the method in the interface, but I haven't seen any code recreating the implementation from scratch rather than extending the base class with empty implementations.

**Comment by peterrehm** [comment]
I have done a rewrite as per @stof's comment in a second commit. It could potentially break if someone has a custom OptionResolver implementation which is i assume unlikely. What do you think now?

File: src/Symfony/Component/Form/ResolvedFormType.php (hunk 7 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/ResolvedFormType.php]: Refactored with the FQN.

**Comment by stof** [comment]
> It could potentially break if someone has a custom OptionResolver implementation which is i assume unlikely.

This is totally unlikely (which is why the interface is deprecated btw) and even more in the case of forms given that the OptionsResolver is instantiated by Symfony

File: src/Symfony/Component/Form/AbstractType.php [hunk 9]
```diff
@@ -44,6 +45,16 @@ public function finishView(FormView $view, FormInterface $form, array $options)
      */
     public function setDefaultOptions(OptionsResolverInterface $resolver)
     {
+        $this->configureOptions($resolver);
+    }
+
+    /**
+     * Configure the options for this type.
```
Comment (stof) [inline src/Symfony/Component/Form/AbstractType.php]: typo: ``Configures``

File: src/Symfony/Component/Form/AbstractTypeExtension.php [hunk 10]
```diff
@@ -44,5 +45,15 @@ public function finishView(FormView $view, FormInterface $form, array $options)
      */
     public function setDefaultOptions(OptionsResolverInterface $resolver)
     {
+        $this->configureOptions($resolver);
+    }
+
+    /**
+     * Configure the options for this type.
```
Comment (stof) [inline src/Symfony/Component/Form/AbstractTypeExtension.php]: typo: missing ``s``

**Comment by peterrehm** [comment]
Yes I just wanted to have it mentioned.

File: UPGRADE-3.0.md [hunk 11]
```diff
@@ -102,6 +102,11 @@ UPGRADE FROM 2.x to 3.0
 
 ### Form
 
+ * The method `AbstractType::setDefaultOptions(OptionsResolverInterface $resolver)` and
+   `AbstractTypeExtension::setDefaultOptions(OptionsResolverInterface $resolver)` have been
+   removed. You should use `AbstractType::configureOptions(OptionsResolver $resolver)` and
+   `AbstractTypeExtension::configureOptions(OptionsResolver $resolver)` instead.
```
Comment (stof) [inline UPGRADE-3.0.md]: the actual change for 3.0 is renaming the method in the *interfaces*

Comment (peterrehm) [inline UPGRADE-3.0.md]: But it is not only renamed, we change the signature as well. Shall i change it?

File: src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php [hunk 12]
```diff
@@ -103,12 +99,51 @@ public function testGetOptionsResolver()
...
+    public function testGetOptionsResolver()
     {
-        if (version_compare(\PHPUnit_Runner_Version::id(), '3.7', '<')) {
-            $this->markTestSkipped('This test requires PHPUnit 3.7.');
-        }
+        $test = $this;
+        $i = 0;
+
+        $assertIndexAndAddOption = function ($index, $option, $default) use (&$i, $test) {
+            return function (OptionsResolverInterface $resolver) use (&$i, $test, $index, $option, $default) {
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php]: please typehint the class rather than the interface

File: src/Symfony/Component/Form/AbstractTypeExtension.php (hunk 10 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/AbstractTypeExtension.php]: Fixed.

File: src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php [hunk 13]
```diff
@@ -103,12 +99,51 @@ public function testGetOptionsResolver()
...
+
+        // The form type itself
+        $this->type->expects($this->once())
+            ->method('configureOptions')
+            ->will($this->returnCallback($assertIndexAndAddOption(1, 'b', 'b_default')));
+
+        // And its extensions
+        $this->extension1->expects($this->once())
+            ->method('configureOptions')
+            ->will($this->returnCallback($assertIndexAndAddOption(2, 'c', 'c_default')));
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php]: ``configureOptions`` is called internally by the form type. It is not called by the ResolvedFormType. So it should not appear here

File: UPGRADE-3.0.md (hunk 11 above)
Comment (stof) [inline UPGRADE-3.0.md]: well, your message shows the typehint. Just keep it when showing the renaming

**Comment by stof** [comment]
Please rebase your branch to fix conflicts. Otherwise Travis is not running the tests

**Comment by peterrehm** [comment]
Rebased. Tests are now showing deprecated warnings which is due to the mock. I am working on this.

**Comment by peterrehm** [comment]
I have refactored tests, now it should be good. This way the tests should work also without adjustments after removing the forward compatibility layer and removing the `setDefaultOptions` method.

**Comment by peterrehm** [comment]
Just added the docs PR. Is all okay now? @stof Are you good with the changes?

**Comment by peterrehm** [comment]
Travis failed on components=high and low but I assume it is not related to that change.

File: src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php [hunk 14]
```diff
@@ -132,10 +127,6 @@ public function testCreateBuilder()
 
     public function testCreateBuilderWithDataClassOption()
     {
-        if (version_compare(\PHPUnit_Runner_Version::id(), '3.7', '<')) {
-            $this->markTestSkipped('This test requires PHPUnit 3.7.');
-        }
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php]: Can you actually do this cleanup in a separate PR on the older branch containing them ?

Comment (peterrehm) [inline src/Symfony/Component/Form/Tests/ResolvedFormTypeTest.php]: @stof I just submitted #13303 with the minor change. Once this is merged into 2.5, 2.6 and 2.7 I can rebase this PR. Anything else to do? /cc @fabpot

**Comment by peterrehm** [comment]
Rebased. Any further comments? Or are we good to merge, @fabpot?

**Comment by peterrehm** [comment]
Travis failed on components=high and low which I don't understand. It looks like other PR's against 2.7 are not failing. Do you have any explanation on this?

**Comment by stof** [comment]
for the components=high, this is because the subtree split of the Form component obviously does not contain the change for configureOptions yet. It should be fixed once the PR is merged and subtrees get updated.

For components=low, this is a combination of 2 things:
- the subtrees for 2.7 are not updated
- the low bound for the symfony/form constraints in other components should be bumped to 2.7 as they need the support for this new method, which is not there in 2.6 and older

**Comment by peterrehm** [comment]
Thanks for the explanation, that makes it clear. Then I am good from my side.

**Comment by peterrehm** [comment]
I have adjusted the version requirements, Is this as you expected it?

**Comment by peterrehm** [comment]
Rebased. Now it looks like just the Tests in the Doctrine and Propel bridge are failing, I assume this is due to the missing subtree split.

File: UPGRADE-3.0.md [hunk 15]
```diff
@@ -106,6 +106,11 @@ UPGRADE FROM 2.x to 3.0
 
 ### Form
 
+ * The method `AbstractType::setDefaultOptions(OptionsResolverInterface $resolver)` and
+   `AbstractTypeExtension::setDefaultOptions(OptionsResolverInterface $resolver)` have been
+   renamed. You should use `AbstractType::configureOptions(OptionsResolver $resolver)` and
```
Comment (WouterJ) [inline UPGRADE-3.0.md]: Isn't the standard to use something like: "have been deprected in favor of ..."

Comment (WouterJ) [inline UPGRADE-3.0.md]: Wait, sorry. this is the upgrade file. Maybe add a before/after example?

Comment (peterrehm) [inline UPGRADE-3.0.md]: @WouterJ I think it is fine as is since before/after is already in the text.

**Comment by peterrehm** [comment]
@fabpot Is there anything blocking this?

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 16]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+2.7.0
+-----
+
+ * deprecated AbstractType::setDefaultOptions() in favor of AbstractType::configureOptions().
```
Comment (stof) [inline src/Symfony/Component/Form/CHANGELOG.md]: actually, what has been deprecated is *overriding* these methods

**Comment by stof** [comment]
The UPGRADE-2.7 file should mention the deprecation and describe the way to upgrade

File: src/Symfony/Component/Form/CHANGELOG.md (hunk 16 above)
Comment (peterrehm) [inline src/Symfony/Component/Form/CHANGELOG.md]: setDefaultOptions() has ben deprecated and the new method exists as preparation for the rename in 3.0.

Comment (stof) [inline src/Symfony/Component/Form/CHANGELOG.md]: no. the method itself is not deprecated (it is not possible as it **must** be implemented in form types and is being used.
What is deprecated is overwriting ``AbstractType::setDefaultOptions()``, in favor of overwriting ``AbstractType::configureOptions()``, to use a forward-compatibility layer.

This is not a BC layer here. We are making it possible to write your code as if you were using the Symfony 3 interface already, even though it is not the case yet, handling the differences in the parent class. This is not the same than providing a new API and deprecating the BC layer

Comment (peterrehm) [inline src/Symfony/Component/Form/CHANGELOG.md]: Well I still consider this as deprecated. Do you want me to change the wording?

**Comment by peterrehm** [comment]
@stof Added the upgrade information

File: src/Symfony/Component/Form/CHANGELOG.md (hunk 16 above)
Comment (stof) [inline src/Symfony/Component/Form/CHANGELOG.md]: I would vote for updating the wording

Comment (peterrehm) [inline src/Symfony/Component/Form/CHANGELOG.md]: @stof updated as per your request.

**Comment by peterrehm** [comment]
@stof Anything else left?

File: src/Symfony/Bridge/Twig/composer.json [hunk 17]
```diff
@@ -22,7 +22,7 @@
     },
     "require-dev": {
         "symfony/finder": "~2.3|~3.0.0",
-        "symfony/form": "~2.6|~3.0.0",
+        "symfony/form": "~2.7|~3.0.0",
```
Comment (stof) [inline src/Symfony/Bridge/Twig/composer.json]: this is not needed actually AFAIK. The Twig bridge does not contain a form type

File: src/Symfony/Bundle/FrameworkBundle/composer.json [hunk 18]
```diff
@@ -39,7 +39,7 @@
         "symfony/finder": "~2.0,>=2.0.5|~3.0.0",
         "symfony/intl": "~2.3|~3.0.0",
         "symfony/security": "~2.6|~3.0.0",
-        "symfony/form": "~2.6|~3.0.0",
+        "symfony/form": "~2.7|~3.0.0",
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: same here

File: src/Symfony/Component/Form/FormTypeInterface.php [hunk 19]
```diff
@@ -72,6 +72,10 @@ public function finishView(FormView $view, FormInterface $form, array $options);
      * Sets the default options for this type.
      *
      * @param OptionsResolverInterface $resolver The resolver for the options.
+     *
+     * @deprecated Deprecated since Symfony 2.7, to be removed in Symfony 3.0.
```
Comment (stof) [inline src/Symfony/Component/Form/FormTypeInterface.php]: ``to be renamed`` would be better

**Comment by peterrehm** [comment]
@stof Updated as per your comments.

**Comment by fabpot** [comment]
@peterrehm Tests do not pass (some problems on Propel seems related to this PR). Also, if not done already, rebase on current master. Thanks.

**Comment by peterrehm** [comment]
@fabpot the only failed tests are on components high and low. @stof mentioned that this is due to a missing subtree split and should be fixed after the merge. I rebased against 2.7, I assume this is what you meant. After this is merge I will update the corresponding PR for the master branch.

**Comment by peterrehm** [comment]
@fabpot So is it okay now?

**Comment by fabpot** [comment]
Thank you @peterrehm.

**Comment by peterrehm** [comment]
Perfect. Once you merge 2.7 into master I will update #13407.
